File: src/lib/api/membership-plans.ts
import { invokeCommand } from "../tauri";

export interface PlanResponse {
  id: string;
  name: string;
  duration_days: number;
  price: number;
  description: string | null;
  is_active: boolean;
  created_at: string;
  updated_at: string;
}

export interface CreatePlanRequest {
  name: string;
  duration_days: number;
  price: number;
  description: string | null;
}

export interface UpdatePlanRequest {
  name: string;
  duration_days: number;
  price: number;
  description: string | null;
}

export async function createPlan(
  request: CreatePlanRequest,
): Promise<PlanResponse> {
  return invokeCommand<PlanResponse>("create_plan", { request });
}

export async function getPlan(id: string): Promise<PlanResponse> {
  return invokeCommand<PlanResponse>("get_plan", { id });
}

export async function listPlans(): Promise<PlanResponse[]> {
  return invokeCommand<PlanResponse[]>("list_plans");
}

export async function listActivePlans(): Promise<PlanResponse[]> {
  return invokeCommand<PlanResponse[]>("list_active_plans");
}

export async function updatePlan(
  id: string,
  request: UpdatePlanRequest,
): Promise<PlanResponse> {
  return invokeCommand<PlanResponse>("update_plan", { id, request });
}

export async function deactivatePlan(id: string): Promise<PlanResponse> {
  return invokeCommand<PlanResponse>("deactivate_plan", { id });
}

File: src-tauri/src/commands/membership_plans.rs
use tauri::State;

use crate::database::Database;
use crate::dto::membership_plan::{CreatePlanRequest, PlanResponse, UpdatePlanRequest};
use crate::errors::AppError;
use crate::services::membership_plan_service;

use super::db::run_db;

#[tauri::command]
pub async fn create_plan(
    state: State<'_, Database>,
    request: CreatePlanRequest,
) -> Result<PlanResponse, AppError> {
    let conn = state.inner().clone_conn();
    run_db(conn, move |c| {
        membership_plan_service::create_plan(c, request)
    })
    .await
}

#[tauri::command]
pub async fn get_plan(state: State<'_, Database>, id: String) -> Result<PlanResponse, AppError> {
    let conn = state.inner().clone_conn();
    run_db(conn, move |c| membership_plan_service::get_plan(c, &id)).await
}

#[tauri::command]
pub async fn list_plans(state: State<'_, Database>) -> Result<Vec<PlanResponse>, AppError> {
    let conn = state.inner().clone_conn();
    run_db(conn, |c| membership_plan_service::list_plans(c)).await
}

#[tauri::command]
pub async fn list_active_plans(state: State<'_, Database>) -> Result<Vec<PlanResponse>, AppError> {
    let conn = state.inner().clone_conn();
    run_db(conn, |c| membership_plan_service::list_active_plans(c)).await
}

#[tauri::command]
pub async fn update_plan(
    state: State<'_, Database>,
    id: String,
    request: UpdatePlanRequest,
) -> Result<PlanResponse, AppError> {
    let conn = state.inner().clone_conn();
    run_db(conn, move |c| {
        membership_plan_service::update_plan(c, &id, request)
    })
    .await
}

#[tauri::command]
pub async fn deactivate_plan(
    state: State<'_, Database>,
    id: String,
) -> Result<PlanResponse, AppError> {
    let conn = state.inner().clone_conn();
    run_db(conn, move |c| {
        membership_plan_service::deactivate_plan(c, &id)
    })
    .await
}

#[tauri::command]
pub async fn reactivate_plan(
    state: State<'_, Database>,
    id: String,
) -> Result<PlanResponse, AppError> {
    let conn = state.inner().clone_conn();
    run_db(conn, move |c| {
        membership_plan_service::reactivate_plan(c, &id)
    })
    .await
}
